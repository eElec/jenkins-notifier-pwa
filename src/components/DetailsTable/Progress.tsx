import React, { useEffect, useState } from 'react';
import { useTheme } from '@mui/system';
import { css } from '@emotion/react';

type Props = { startTime: number; estDuration: number };

function msToTime(s: number) {
	// Pad to 2 or 3 digits, default is 2
	var pad = (n: number, z = 2) => ('00' + n).slice(-z);
	return (
		pad((s / 3.6e6) | 0) +
		':' +
		pad(((s % 3.6e6) / 6e4) | 0) +
		':' +
		pad(((s % 6e4) / 1000) | 0)
	);
}

function Progress(props: Props) {
	const { startTime, estDuration } = props;
	const theme = useTheme();

	// Update every second
	const [currentTime, setCurrentTime] = useState(Date.now());

	useEffect(() => {
		const interval = setInterval(() => {
			setCurrentTime(Date.now());
		}, 1000);
		return () => {
			clearInterval(interval);
		};
	});

	const perc = Math.min(((currentTime - startTime) / estDuration) * 100, 100);

	const startString = new Date(startTime).toTimeString().substring(0, 8);
	const endString = new Date(startTime + estDuration)
		.toTimeString()
		.substring(0, 8);

	return (
		<div
			id="progress"
			css={{
				position: 'absolute',
				bottom: 0,
				left: 0,
				height: '2px',
				width: '100%',
				display: 'flex',
				justifyContent: 'space-between',
				fontSize: '0px',
				span: {
					position: 'relative',
					bottom: '1px',
					color: theme.palette.text.primary,
				},
				transition: 'ease 0.2s',
			}}
		>
			<div
				css={{
					position: 'absolute',
					bottom: 0,
					left: 0,
					background:
						theme.palette.mode === 'light'
							? theme.palette.success.light
							: theme.palette.success.dark,
					height: '100%',
				}}
				style={{ width: `${perc}%` }}
			/>
			<span>{startString}</span>
			<span>{msToTime(currentTime - startTime)}</span>
			<span>{endString}</span>
		</div>
	);
}

export default Progress;
