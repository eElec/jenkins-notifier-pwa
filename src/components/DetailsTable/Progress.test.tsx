import { describe, expect, it, vi } from 'vitest';
import { render, screen, act } from '@test/util';
import Progress from './Progress';

describe.only('Progress Component', () => {
	const startTime = new Date(2022, 1, 1);
	const estDuration = 1000 * 60 * 60; // 1 hr

	beforeEach(() => {
		vi.useFakeTimers();
		vi.setSystemTime(startTime);
	});
	afterEach(() => vi.useRealTimers());

	it('renders', () => {
		const result = render(
			<Progress startTime={startTime.getTime()} estDuration={estDuration} />
		).baseElement;
		expect(result).toMatchSnapshot();
	});

	describe('static', () => {
		it('0%', () => {
			const result = render(
				<Progress startTime={startTime.getTime()} estDuration={estDuration} />
			);

			expect(screen.getByTestId('progress-start').textContent).toEqual(
				'00:00:00'
			);
			expect(screen.getByTestId('progress-elapsed').textContent).toEqual(
				'00:00:00'
			);
			expect(screen.getByTestId('progress-end').textContent).toEqual(
				'01:00:00'
			);

			expect(screen.getByTestId('progress-bar').style.width).toEqual('0%');
		});

		it('50%', () => {
			const result = render(
				<Progress startTime={startTime.getTime()} estDuration={estDuration} />
			);

			act(() => {
				vi.advanceTimersByTime(estDuration / 2); // Advance by 30min
			});

			expect(screen.getByTestId('progress-start').textContent).toEqual(
				'00:00:00'
			);
			expect(screen.getByTestId('progress-elapsed').textContent).toEqual(
				'00:30:00'
			);
			expect(screen.getByTestId('progress-end').textContent).toEqual(
				'01:00:00'
			);

			expect(screen.getByTestId('progress-bar').style.width).toEqual('50%');
		});

		it('100%', () => {
			const result = render(
				<Progress startTime={startTime.getTime()} estDuration={estDuration} />
			);

			act(() => {
				vi.advanceTimersByTime(estDuration); // Advance by 30min
			});

			expect(screen.getByTestId('progress-start').textContent).toEqual(
				'00:00:00'
			);
			expect(screen.getByTestId('progress-elapsed').textContent).toEqual(
				'01:00:00'
			);
			expect(screen.getByTestId('progress-end').textContent).toEqual(
				'01:00:00'
			);

			expect(screen.getByTestId('progress-bar').style.width).toEqual('100%');
		});

		it('Elapsed time exceeds estimated duration', () => {
			const result = render(
				<Progress startTime={startTime.getTime()} estDuration={estDuration} />
			);

			act(() => {
				vi.advanceTimersByTime(estDuration * 1.5); // Advance by 30min
			});

			expect(screen.getByTestId('progress-start').textContent).toEqual(
				'00:00:00'
			);
			expect(screen.getByTestId('progress-elapsed').textContent).toEqual(
				'01:30:00'
			);
			expect(screen.getByTestId('progress-end').textContent).toEqual(
				'01:00:00'
			);

			expect(screen.getByTestId('progress-bar').style.width).toEqual('100%');
		});
	});

	it.only('timer removed on unmount', () => {
		expect(vi.getTimerCount()).toBe(0);
		const result = render(
			<Progress startTime={startTime.getTime()} estDuration={estDuration} />
		);
		expect(vi.getTimerCount()).toBe(1);
		result.unmount();
		expect(vi.getTimerCount()).toBe(0);
	});
});
