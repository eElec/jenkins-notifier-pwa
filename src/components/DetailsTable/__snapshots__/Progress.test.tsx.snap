// Vitest Snapshot v1

exports[`Progress Component > renders 1`] = `
.emotion-0 {
  position: absolute;
  bottom: 0;
  left: 0;
  height: 2px;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  font-size: 0px;
  -webkit-transition: ease 0.2s;
  transition: ease 0.2s;
}

.emotion-0 span {
  position: relative;
  bottom: 1px;
  color: #fff;
}

.emotion-1 {
  position: absolute;
  bottom: 0;
  left: 0;
  background: rgb(88, 175, 78);
  height: 100%;
}

<body>
  <div>
    <div
      class="emotion-0"
      id="progress"
    >
      <div
        class="emotion-1"
        data-testid="progress-bar"
        style="width: 0%;"
      />
      <span
        data-testid="progress-start"
      >
        00:00:00
      </span>
      <span
        data-testid="progress-elapsed"
      >
        00:00:00
      </span>
      <span
        data-testid="progress-end"
      >
        01:00:00
      </span>
    </div>
  </div>
</body>
`;
